src/Stratis.Bitcoin/P2P/PeerSelector.cs

```diff
@@ -41,30 +43,241 @@ public sealed class PeerSelector : IPeerSelector
         }
 
         /// <inheritdoc/>
-        public IEnumerable<PeerAddress> SelectPeers()
+        public IEnumerable<PeerAddress> SelectPeers(int amountOfPeers)
+        {
+            // If there are no peers, just one or if the amount peers is less than
+            // the amount of peers asked for, just return the list.
+            if (!this.peerAddresses.Any() || this.peerAddresses.Count == 1 || this.peerAddresses.Count < amountOfPeers)
+                return this.peerAddresses.Select(pa => pa.Value);
+
+            // If there are more peers than what we asked for, apply a fill ratio.
+            var random = new Random();
+
+            // Randomly order the set of handshaked peers.
+            var handshaked = this.peerAddresses.Handshaked().OrderBy(p => random.Next());
+
+            // Randomly order the set of connected peers.
+            var connected = this.peerAddresses.Connected().OrderBy(p => random.Next());
+
+            // Randomly order the set of attempted and fresh peers.
+            var attemptedAndFresh = this.peerAddresses.Attempted().Concat(this.peerAddresses.Fresh()).OrderBy(p => random.Next());
+
+            // If handshaked, connected and attempted or fresh peers exist, apply the
+            // default ratio.
+            if (handshaked.Any() && connected.Any() && attemptedAndFresh.Any())
+                return FillWithDefaultRatios(amountOfPeers, handshaked, connected, attemptedAndFresh);
+
+            // If only 2 sets exist, apply the secondary ratio.
+            if (!handshaked.Any() && connected.Any() && attemptedAndFresh.Any())
+            {
+                return FillWithSecondaryRatios(amountOfPeers, connected, attemptedAndFresh);
+            }
+
+            // If only 2 sets exist, apply the secondary ratio.
+            if (handshaked.Any() && !connected.Any() && attemptedAndFresh.Any())
+            {
+                return FillWithSecondaryRatios(amountOfPeers, handshaked, attemptedAndFresh);
+            }
+
+            // If only 2 sets exist, apply the secondary ratio.
+            if (handshaked.Any() && connected.Any() && !attemptedAndFresh.Any())
+            {
+                return FillWithSecondaryRatios(amountOfPeers, handshaked, connected);
+            }
+
+            // If there is only 1 set, take the amount of peers requested.
+            if (!handshaked.Any() && !connected.Any() && attemptedAndFresh.Any())
+            {
+                return attemptedAndFresh.Take(amountOfPeers);
+            }
+
+            // If there is only 1 set, take the amount of peers requested.
+            if (handshaked.Any() && !connected.Any() && !attemptedAndFresh.Any())
+            {
+                return handshaked.Take(amountOfPeers);
+            }
+
+            // If there is only 1 set, take the amount of peers requested.
+            if (!handshaked.Any() && connected.Any() && !attemptedAndFresh.Any())
+            {
+                return connected.Take(amountOfPeers);
```

Comment:
this if series is hardly readable, can we improve it

Reply:
yes I agree it could be beteter

Reply:
@dangershony This is an old commit ;) This logic has been removed 👍